### PR TITLE
Add tests for complex number type support in algebras

### PR DESCRIPTION
**Pull request by hugohadfield**
Adds tests to ensure that complex number type support behaves as one would expect and type info is not lost, starts addressing https://github.com/pygae/clifford/issues/354

### CODE REVIEW

**Comment by hugohadfield** [comment]
My knowledge of pytest.fixture is not very good, @eric-wieser any suggestions to get these tests to run over the conformalised metric as well?

File: clifford/test/test_complex.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        res = (A + 1j*B).value
+        res2 = A.value + 1j*B.value
+        np.testing.assert_allclose(res, res2)
+
+    def test_subtraction(self, algebra):
+        A = algebra.randomMV()
+        B = algebra.randomMV()
+        res = (A - 1j*B).value
+        res2 = A.value - 1j*B.value
+        np.testing.assert_allclose(res, res2)
```
Comment (eric-wieser) [inline clifford/test/test_complex.py:57]: I think for these two tests you should require exact equality

File: clifford/test/test_complex.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+    def test_ip(self, algebra):
+        A = algebra.randomMV()
+        B = algebra.randomMV()
+        C = algebra.randomMV()
+        D = algebra.randomMV()
+        res = ((A + 1j*B)|(C + 1j*D)).value
+        res2 = (A|C).value + 1j*(B|C).value + 1j*(A|D).value - (B|D).value 
+        np.testing.assert_allclose(res, res2)
```
Comment (eric-wieser) [inline clifford/test/test_complex.py:84]: Parametrization can merge these:
```suggestion
    @pytest.mark.parametrize('p', [cf.operator.gp, cf.operator.op, cf.operator.ip])
    def test_prod(self, algebra, p):
        A = algebra.randomMV()
        B = algebra.randomMV()
        C = algebra.randomMV()
        D = algebra.randomMV()
        res = (p(A + 1j*B, C + 1j*D)).value
        res2 = p(A, C).value + 1j*p(B, C).value + 1j*p(A, D).value - p(B, D).value 
        np.testing.assert_allclose(res, res2)
```
(with appropriate `import clifford as cf` or similar)

File: clifford/test/test_complex.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        res = (~(A + 1j*B)).value
+        res2 = (~A).value + 1j*(~B).value
+        np.testing.assert_allclose(res, res2)
+
+    def test_grade_selection(self, algebra):
+        A = algebra.randomMV()
+        B = algebra.randomMV()
+        res = ((A + 1j*B)(2)).value
+        res2 = A(2).value + 1j*B(2).value
+        np.testing.assert_allclose(res, res2)
```
Comment (eric-wieser) [inline clifford/test/test_complex.py:98]: Again, you should require exact equality here.

File: clifford/test/test_complex.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+
```
Comment (eric-wieser) [inline clifford/test/test_complex.py:100]:
```suggestion
```

File: clifford/test/test_complex.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+@pytest.fixture(scope='module')
+def g4_1():
+    return Cl(4, 1)[0]
+
+
+@pytest.fixture(scope='module')
+def g3c():
+    return conformalize(Cl(3)[0])[0]
```
Comment (eric-wieser) [inline clifford/test/test_complex.py:37]: Using both seems unnecessary, given you're not actually using any of the other methods.

File: clifford/test/test_complex.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
+
```
Comment (eric-wieser) [inline clifford/test/test_complex.py:1]: A docstring here would be good - perhaps something like

> the tests below are based on an informed guess about the correct behavior, and we should not be afraid of changing them if the things we're testing for turn out to not be the conventional definitions

File: clifford/_multivector.py [hunk 7]
```diff
@@ -496,10 +496,10 @@ def __str__(self) -> str:
             else:
                 if coeff < 0:
                     sep = seps[1]
-                    abs_coeff = -round(coeff, p)
+                    abs_coeff = -np.round(coeff, p)
                 else:
                     sep = seps[0]
-                    abs_coeff = round(coeff, p)
+                    abs_coeff = np.round(coeff, p)
```
Comment (eric-wieser) [inline clifford/_multivector.py:502]: This seems risky - it causes a crash for `str(2**64 * e1)`.

Comment (hugohadfield) [reply clifford/_multivector.py:502]: So I did this to fix the deprecation warning:
```
clifford/test/test_complex.py: 53 warnings
  C:\Users\hughad01\internship\clifford\clifford\_multivector.py:499: DeprecationWarning: The Python built-in `round` is deprecated for complex scalars, and will raise a `TypeError` in a future release. Use `np.round` or `scalar.round` instead.
    abs_coeff = -round(coeff, p)
```
Should we use the scalar.round option instead?

Comment (eric-wieser) [reply clifford/_multivector.py:502]: That doesn't exist for the case I give above either

**Comment by hugohadfield** [comment]
Travis failures appear to be for unrelated reasons, squash and merge?

**Comment by eric-wieser** [comment]
3.5 CI failure is caused by https://github.com/davidhalter/parso/issues/149

**Comment by eric-wieser** [comment]
Other failures indeed look unrelated, although I'm not sure when they crept in

File: clifford/_layout.py [hunk 8]
```diff
@@ -574,7 +574,7 @@ def _gen_complement_func(self, omt):
 
         @_numba_utils.njit
         def comp_func(Xval):
-            Yval = np.zeros(dims)
+            Yval = np.zeros(dims, dtype=Xval.dtype)
```
Comment (eric-wieser) [inline clifford/_layout.py:577]: Can you add a test to `test_clifford::TestClifford.test_unary_op_preserves_dtype` for this?

File: clifford/_multivector.py (hunk 7 above)
Comment (eric-wieser) [reply clifford/_multivector.py:502]: Can we either back out this change, or add a `if np.issubdtype(self.value.dtype, np.number` around the use of `np.round`?

File: clifford/_layout.py [hunk 9]
```diff
@@ -447,7 +447,7 @@ def dual_func(self):
         else:
             # Equivalent to but faster than
             #   Iinv = self.pseudoScalar.inv().value
-            Iinv = np.zeros(self.gaDims)
+            Iinv = np.zeros(self.gaDims, dtype=int)
```
Comment (hugohadfield) [inline clifford/_layout.py:450]: @eric-wieser do you think this is a safe thing to do? I was aiming for a type that could be upgraded to whatever the type of the multivector was

Comment (eric-wieser) [reply clifford/_layout.py:450]: Ideally you'd pick the type based on the type of `1 / II_scalar`

Comment (eric-wieser) [reply clifford/_layout.py:450]: Otherwise this will go wrong for metrics containing 2

File: clifford/test/test_clifford.py [hunk 10]
```diff
@@ -386,6 +386,8 @@ def test_binary_op_preserves_dtype(self, dtype, func, g3):
         operator.pos,
         operator.neg,
         MultiVector.gradeInvol,
+        MultiVector.dual,
+        MultiVector.right_complement
```
Comment (eric-wieser) [inline clifford/test/test_clifford.py:390]:
```suggestion
        MultiVector.right_complement,
        MultiVector.left_complement,
```

File: clifford/_layout.py [hunk 11]
```diff
@@ -447,10 +447,14 @@ def dual_func(self):
         else:
             # Equivalent to but faster than
             #   Iinv = self.pseudoScalar.inv().value
-            Iinv = np.zeros(self.gaDims)
             II_scalar = self.gmt[-1, 0, -1]
+            inv_II_scalar = 1 / II_scalar
+            if II_scalar == 1 or -1:
```
Comment (eric-wieser) [inline clifford/_layout.py:452]:
```suggestion
            if II_scalar in (1, -1):
```

Comment (eric-wieser) [reply clifford/_layout.py:452]: What you had was nonsense

File: clifford/test/test_clifford.py [hunk 12]
```diff
@@ -386,6 +386,8 @@ def test_binary_op_preserves_dtype(self, dtype, func, g3):
         operator.pos,
```
Comment (eric-wieser) [inline clifford/test/test_clifford.py:386]: I think you should be able to add `operator.and_,  # vee product` on line 375 now too.

File: clifford/_multivector.py [hunk 13]
```diff
@@ -118,12 +118,42 @@ def vee(self, other) -> 'MultiVector':
...
+        other, mv = self._checkOther(other, coerce=True)
+
+        if mv:
+            newValue = self.layout.vee_func(self.value, other.value)
+        else:
+            if isinstance(other, np.ndarray):
+                obj = self.__array__()
+                return obj&other
+
+            newValue = self.value&other
```
Comment (eric-wieser) [inline clifford/_multivector.py:130]: This line is wrong, it performs bitwise and. This code-path is taken by `(5*e1) & 3`, and should give the same result as if `3` were first turned into a real multivector.

File: clifford/_multivector.py [hunk 14]
```diff
@@ -118,12 +118,42 @@ def vee(self, other) -> 'MultiVector':
...
+
+        return self.layout.MultiVector(newValue)
+
+    def _rvee(self, other) -> 'MultiVector':
```
Comment (eric-wieser) [inline clifford/_multivector.py:134]: Hmm, this is gross.

File: clifford/_multivector.py [hunk 15]
```diff
@@ -118,12 +118,42 @@ def vee(self, other) -> 'MultiVector':
...
         """
-        return self.layout.MultiVector(value=self.layout.vee_func(self.value, other.value))
+        other, mv = self._checkOther(other, coerce=True)
+
+        if mv:
+            newValue = self.layout.vee_func(self.value, other.value)
+        else:
+            if isinstance(other, np.ndarray):
+                obj = self.__array__()
+                return obj&other
```
Comment (eric-wieser) [inline clifford/_multivector.py:128]: (note this line is fine, because here `obj` is a MVarray)

File: clifford/_multivector.py (hunk 13 above)
Comment (hugohadfield) [reply clifford/_multivector.py:130]: Surely in that case 3 would have been coerced into a `MultiVector` during `checkother(3, coerce=True)`?

Comment (eric-wieser) [reply clifford/_multivector.py:130]: I missed the coerce=True. I'm not certain this line is reachable - so I'd be inclined to raise `TypeError`.

It's pretty obvious that our coercion rules needs some work. I'll maybe revisit this in another PR.

**Review by eric-wieser — APPROVED** [review]

**Pull request merged**